File: LintCode/ 88. Lowest Common Ancestor of a Binary Tree.py
# https://www.lintcode.com/problem/lowest-common-ancestor-of-a-binary-tree/description
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""


class Solution:
    """
    @param: root: The root of the binary search tree.
    @param: A: A TreeNode in a Binary.
    @param: B: A TreeNode in a Binary.
    @return: Return the least common ancestor(LCA) of the two nodes.
    """
    def lowestCommonAncestor(self, root, A, B):
        if root is None:
            return None
            
        if A == root or B == root:
            return root
            
        result_left = self.lowestCommonAncestor(root.left, A, B)
        result_right = self.lowestCommonAncestor(root.right,A, B)
        
        if result_left and result_right:
            return root
        
        if result_left:
            return result_left
        
        if result_right:
            return result_right
            
        return None

File: lintcode1/Cpp/88.LowestCommonAncestorofaBinaryTree.cpp
// https://www.lintcode.com/problem/lowest-common-ancestor-of-a-binary-tree/my-submissions
/* recursion */

/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */


class Solution {
public:
    /*
     * @param root: The root of the binary search tree.
     * @param A: A TreeNode in a Binary.
     * @param B: A TreeNode in a Binary.
     * @return: Return the least common ancestor(LCA) of the two nodes.
     */
    TreeNode * lowestCommonAncestor(TreeNode * root, TreeNode * A, TreeNode * B) {
        // write your code here
        if(NULL == root)
        {
            return NULL;
        }
        
        if(A == root || B == root)
        {
            return root;
        }
        
        TreeNode * left = lowestCommonAncestor(root->left,A,B);
        TreeNode * right = lowestCommonAncestor(root->right,A,B);
        
        if(NULL != left && NULL != right){
            return root;
        }
        
        if(NULL != left)
        {
            return left;
        }
        
        if(NULL != right)
        {
            return right;
        }
        
        return NULL;
    }
};